### PR TITLE
[KYUUBI #670] KyuubiSQLException support sqlState and errorCode

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

#670

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala [hunk 1]
```diff
@@ -42,27 +55,28 @@ class KyuubiSQLException(msg: String, cause: Throwable) extends SQLException(msg
   }
 }
 
-object KyuubiSQLException {
+object KyuubiSQLException extends Logging {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:58]: rm Logging?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiException.scala [hunk 2]
```diff
@@ -17,7 +17,4 @@
 
 package org.apache.kyuubi
 
-class KyuubiException(message: String, cause: Throwable) extends Exception(message, cause) {
-
-  def this(message: String) = this(message, null)
-}
+class KyuubiException(message: String, cause: Throwable = null) extends Exception(message, cause)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiException.scala:20]: seems out of the PR scope

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiException.scala]: reverted.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala (hunk 1 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala]: updated.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala [hunk 3]
```diff
@@ -26,7 +26,20 @@ import scala.collection.JavaConverters._
...
-class KyuubiSQLException(msg: String, cause: Throwable) extends SQLException(msg, cause) {
+/**
+ * @param msg        a description of the exception
+ * @param sqlState   an XOPEN or SQL:2003 code identifying the exception
+ * @param errorCode  a database vendor-specific exception code
+ * @param cause      the underlying reason for this [[SQLException]]
+ *                   (which is saved for later retrieval by the `getCause()` method);
+ *                   may be null indicating the cause is non-existent or unknown.
+ */
+class KyuubiSQLException(msg: String, cause: Throwable, sqlState: String, errorCode: Int)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:37]: the constructor paras order is not the same with SQLException and the comment

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala [hunk 4]
```diff
@@ -47,22 +60,23 @@ object KyuubiSQLException {
   private final val HEAD_MARK: String = "*"
   private final val SEPARATOR: Char = ':'
 
-  def apply(msg: String, throwable: Throwable): KyuubiSQLException = {
-    new KyuubiSQLException(msg, findCause(throwable))
+  def apply(msg: String,
+            cause: Throwable = null,
+            sqlState: String = null,
+            errorCode: Int = 0): KyuubiSQLException = {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:66]: maybe venderCode is more accurate?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala [hunk 5]
```diff
@@ -26,7 +26,20 @@ import scala.collection.JavaConverters._
...
+ * @param sqlState   an XOPEN or SQL:2003 code identifying the exception
+ * @param errorCode  a database vendor-specific exception code
+ * @param cause      the underlying reason for this [[SQLException]]
+ *                   (which is saved for later retrieval by the `getCause()` method);
+ *                   may be null indicating the cause is non-existent or unknown.
+ */
+class KyuubiSQLException(msg: String, cause: Throwable, sqlState: String, errorCode: Int)
+  extends SQLException(msg, sqlState, errorCode, cause) {
+
+  // for reflection
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:40]: where do we use this?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala (hunk 4 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:66]: ident

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala (hunk 5 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:40]: https://github.com/NetEase/kyuubi/blob/b9079b698d0239adad1fda945cb82f59c7eca950/kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala#L116-L124

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala (hunk 3 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:37]: will align SQLException

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala [hunk 6]
```diff
@@ -47,22 +60,24 @@ object KyuubiSQLException {
   private final val HEAD_MARK: String = "*"
   private final val SEPARATOR: Char = ':'
 
-  def apply(msg: String, throwable: Throwable): KyuubiSQLException = {
-    new KyuubiSQLException(msg, findCause(throwable))
+  def apply(
+      msg: String,
+      cause: Throwable = null,
+      sqlState: String = null,
+      vendorCode: Int = 0): KyuubiSQLException = {
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:67]: reorder the parameters to simplify method overwrite

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala (hunk 6 above)
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:67]: `vendorCode: Int = 0`

Why we use `0` as the default code, seems `0` is success. use `-1` ?

Comment (ulysses-you) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/KyuubiSQLException.scala:67]: nvm, `SQLException` also use 0 ..

**Comment by ulysses-you** [comment]
thanks, merging to master for v1.3.0

**Pull request closed without merging**